pkg/syncer/metadatasyncer.go

```diff
@@ -831,6 +839,318 @@ func InitMetadataSyncer(ctx context.Context, clusterFlavor cnstypes.CnsClusterFl
+}
+
+// calculatePendingPVCReservedForNamespace calculate the expected reserved value for PVC for namespace
+func calculatePVCReservedForNamespace(ctx context.Context, cnsOperatorClient client.Client,
+	volumeMap map[string]v1.PersistentVolume, namespace string) (map[string]*resource.Quantity, error) {
+	log := logger.GetLogger(ctx)
+	scTospuMap, err := fetchStorageClassToStoragePolicyMapping(ctx, cnsOperatorClient)
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("Fetching PersistentVolumeClaim for namespace %q", namespace)
+	pvcList := &v1.PersistentVolumeClaimList{}
+	err = cnsOperatorClient.List(ctx, pvcList, &client.ListOptions{Namespace: namespace})
+	if err != nil {
+		return nil, err
+	}
+	spuWiseReservedMap := make(map[string]*resource.Quantity)
+	for _, pvc := range pvcList.Items {
+		log.Infof("Processing PersistentVolumeClaim Name: %q, Namespace: %q", pvc.Name, pvc.Namespace)
+		if pvc.DeletionTimestamp != nil {
+			log.Infof("pvc is marked for deletion, ignoring Name: %q, Namespace: %q",
+				pvc.Name, pvc.Namespace)
+			continue
+		}
+		if pvc.Status.Phase == "" {
+			log.Infof("pvc status not populated, continuing processing others Name: %q, Namespace: %q",
+				pvc.Name, pvc.Namespace)
+			continue
+		}
+		if spu, ok := scTospuMap[*pvc.Spec.StorageClassName]; ok {
+			if spuWiseReservedMap[spu] == nil {
+				spuWiseReservedMap[spu] = resource.NewQuantity(0, resource.BinarySI)
+			}
+		}
+		if pvc.Status.Phase == v1.ClaimPending {
+			if _, ok := scTospuMap[*pvc.Spec.StorageClassName]; ok {
+				log.Infof("pvc is in pending state, Name: %q, Namespace: %q adding pvc capacity to reserved",
+					pvc.Name, pvc.Namespace)
+				spuWiseReservedMap[*pvc.Spec.StorageClassName].Add(
+					pvc.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)])
+			}
+		} else if pv, ok := volumeMap[pvc.Name]; ok {
+			if pv.Spec.Capacity[v1.ResourceName(v1.ResourceStorage)] !=
+				pvc.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)] {
+				log.Infof("pvc size being expanded Name: %q, Namespace: %q adding pvc capacity to reserved",
+					pvc.Name, pvc.Namespace)
+				if _, ok := scTospuMap[*pvc.Spec.StorageClassName]; ok {
+					spuWiseReservedMap[*pvc.Spec.StorageClassName].Add(pvc.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)])
+				}
+			}
+		}
+	}
+	return spuWiseReservedMap, nil
+}
+
+// calculateVolumeSnapshotReservedForNamespace calculate the expected reserved value for volumesnapshot for namespace
+func calculateVolumeSnapshotReservedForNamespace(ctx context.Context, cnsOperatorClient client.Client,
+	namespace string) (map[string]*resource.Quantity, error) {
+	log := logger.GetLogger(ctx)
+	scTospuMap, err := fetchStorageClassToStoragePolicyMapping(ctx, cnsOperatorClient)
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("Fetching VolumeSnapshots for namespace %q", namespace)
+	vsList := &snap.VolumeSnapshotList{}
+	err = cnsOperatorClient.List(ctx, vsList, &client.ListOptions{Namespace: namespace})
```

Comment:
Same, create a lister for volumesnapshots, this way you can sure this thing executes fast